File: 53-maximum-subarray/53-maximum-subarray.java
class Solution {
    public int maxSubArray(int[] nums) {
        int n = nums.length;
        int maxSum = Integer.MIN_VALUE;
        int sum = 0;
        for(int i=0; i<n; i++){
            sum += nums[i];
            if(sum > maxSum){
                maxSum = sum;
            }
            if(sum < 0){
                sum = 0;
            }
        }
        return maxSum;
    }
}

File: 53-maximum-subarray/53-maximum-subarray.cpp
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        int maxSum = INT_MIN;
        int currSum = 0;
        for(int i=0; i<nums.size(); i++){
            currSum += nums[i];
            if(currSum > maxSum){
                maxSum = currSum;
            }
            if(currSum <= 0){
                // if(nums[i] > maxSum){
                //     maxSum = nums[i];
                //     currSum = nums[i];
                // }
                // else{
                //     currSum = 0;
                // }
                currSum = max(0, nums[i]);
                maxSum = max(maxSum, nums[i]);
            }
        }
        return maxSum;
    }
};